Added j!assets and j!assets richest [value || amount]

E1337Kat opened the pull request:
Users can check on just their assets in their possession, as well as rank the server by asset value or total number of shares owned.

E1337Kat committed:
Adding leaderboard by stock prices.

E1337Kat committed:
Fucking doing this, richest can order by assets held or assets value

E1337Kat committed:
":hammer_pick: The revolution comes and your wealth of `§{}` is redistributed." --> ":hammer_pick: The revolution comes and requires `§{}` to fund the new order."

E1337Kat commented:
I have not been able to run this on any server yet, as it is too late at night... maybe tomorrow.

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -175,6 +175,94 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+    @commands.group(pass_context=True, invoke_without_command=True, aliases=["shares", "share", "asset"])
+    async def assests(self, ctx: Context, *, target: discord.Member = None):
+        """
+        Gets the current amount of § a user has in assests.
```
This isn't right

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -175,6 +175,94 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
         await self.bot.redis.increase_history_count(message.channel)
 
+    @commands.group(pass_context=True, invoke_without_command=True, aliases=["shares", "share", "asset"])
+    async def assests(self, ctx: Context, *, target: discord.Member = None):
```
Typo (should be `assets`)

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -175,6 +175,94 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+        """
+        Gets the current amount of § a user has in assests.
+
+        If no target is provided, it will show your balance.
```
Also not right, please update the docstring

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -175,6 +175,94 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+        """
+        user = target or ctx.message.author
+        if user.bot:
+            await ctx.channel.send(":x: Bots cannot earn money.")
```
Should be `cannot own shares`

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -175,6 +175,94 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+            em.colour = user.colour
+            await ctx.send(embed=em)
+        else:
+            await ctx.channel.send("User **{}** has `§{}`.".format(user, currency))
```
This isn't right for this command

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -175,6 +175,94 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+
+            ts = datetime.datetime(year=2008, month=7, day=day, hour=hour, minute=minute, second=second)
+
+            em = discord.Embed(title="First National Bank of Joku")
```
Tis the 1st Stock Market

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -175,6 +175,94 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+    @assets.command(pass_context=True, aliases=["top", "leaderboard"])
+    async def richest(self, ctx: Context, *, what: str = "value"):
+        """
+        Shows the top 10 stock holders in this guild by asset worth
```
`.`

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -175,6 +175,94 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+            return
+        if what == "worth": #order by amount of stock owned
+            users = await ctx.bot.database.get_multiple_users(*ctx.message.guild.members, 
+                                                           order_by=sum(userstock.amount * userstock.stock.price for 
```
I don't think this will actually work, if I know SQLAlchemy.

Also, `stocks` would be undefined.

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -507,6 +595,8 @@ def _get_relative(val) -> typing.Tuple[str, float]:
         await ctx.channel.send(file=buf, filename="plot.png")
 
+
+
```
what is this

Fuyukai reviewed (changes requested):
I guess I could help remove some of the code dupe but fix the issues first please.

E1337Kat committed:
Merge branch 'master' of https://github.com/SunDwarf/Jokusoramame

E1337Kat committed:
fixing errors, need to rework sort

E1337Kat committed:
reworked database calls and sort by amount... it should work.. maybe. I'm not so good at this

E1337Kat committed:
PEP 8

E1337Kat commented:
Fixed those errors, reworked the database calls... I *think* it should work, but again, I am not familiar with postgresql nor python(learning it quickly tho). I was trying to test it on my machine, but it seems my laptop does not want to find the ruamel.yaml stuff... even though it is just right fricking in the folder.

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -175,6 +175,97 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+            await ctx.send(":x: Stocks are not enabled for this server.")
+            return
+
+        #Display only Assets the user owns (stolen from currency)
```
PEP 8 THAT SHIT UP

(needs a space between `#` and `D`)

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -175,6 +175,97 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+            except AttributeError:
+                # Prevent race condition - member leaving between command invocation and here
+                continue
+            stocks = await ctx.bot.database.get_user_stocks(u)
```
why do you load this twice?

Fuyukai replied:
for the record, I try to avoid as many DB calls as possible for performance reasons.

E1337Kat replied:
So I tried not to call the DB as much as I could, but it doesn't seem like there is another way to get it to work. I'll take another look at it though and see if I can get it done with fewer DB calls. I'd like it to be a quicker command as well.

E1337Kat replied:
The first call to get the stocks was to help sort the list, the second call was to get and use the stocks... Maybe I could append a column onto the list...

Fuyukai reviewed (commented):
Looks better, just fix this and I'll rebase onto master.

E1337Kat committed:
Merge https://github.com/SunDwarf/Jokusoramame

E1337Kat committed:
Reworked logic to reduce database calls. Appends columns to users list, then sorts.

E1337Kat committed:
Merge branch 'master' of https://github.com/e1337kat/Jokusoramame

E1337Kat commented:
Everything should be good now. Let me know how to run my own version of the bot, that way I can test it myself in the future.

E1337Kat committed:
PEP 8git push git push

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -176,6 +176,103 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+        for user in users:        
+            stocks = await ctx.bot.database.get_user_stocks(user)
+            # Append a column amount to list users
+            user + [sum(userstock.amount for userstock in stocks]
```
This is a no-op - the result of the addition isn't actually stored anywhere.

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -176,6 +176,103 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+            # Append a column amount to list users
+            user + [sum(userstock.amount for userstock in stocks]
+            # Append a column worth to list users
+            user + [sum(userstock.amount * userstock.stock.price for userstock in stocks)]
```
Same here.

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -176,6 +176,103 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+            # Append a column worth to list users
+            user + [sum(userstock.amount * userstock.stock.price for userstock in stocks)]
+                
+        if what == "amount" or "owned": #order by stock amount
```
This doesn't evaluate to what you think it does:

```
>>> what == "amount" or "owned"
'owned'
```
This actually means `(what == "amount") or "owned"`, and since `"owned"` is a non-zero length string it will always evaluate to True if `what` isn't `"amount"`.

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -176,6 +176,103 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+        # Create a table using tabulate.
+        if what == "owned":
+            headers = ["POS", "User", "Assets Owned"]
+        else # what = default: order by value of stock worth
```
Missing a `:`.

Fuyukai reviewed (changes requested):
I'll add a guide to setting up a development environment in a sec, so you can actually test the code first >.>

E1337Kat committed:
simple changes as suggested

E1337Kat committed:
Merge commit '9f5a9ed89f4e5b9c9b3c69876ab77ec10e3487f7'

E1337Kat committed:
Fixed syntax error

Fuyukai commented on joku/cogs/games/stocks.py:
```diff
@@ -235,21 +235,21 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+            user = [user + [sum(userstock.amount * userstock.stock.price for userstock in stocks)]]
                 
-        if what == "amount" or "owned": #order by stock amount
+        if what == ("amount" or "owned"): #order by stock amount
```
```
>>> what = "amount"
>>> what == ("amount" or "owned")
True
>>> what = "owned"
>>> what == ("amount" or "owned")
False
```

E1337Kat replied:
God damn I'm a fucking idiot... I've written the correct code so many fucking times, I have no idea how fucked that up so hard. bleh. 

if ((what == "amount") or (what == "owned")): 

I should quit programming and give up on life.

Fuyukai replied:
Alternate form: `if what in ("amount", "owned"):`

E1337Kat replied:
oh cool! That makes sense as well... I'm starting to like python more and more. I will probably be leaving it as what I already changed it to unless i do have to change more code tode(which seems likely)

E1337Kat committed:
I'm a fucking idiot.... fixed conditionals

E1337Kat committed:
Merge https://github.com/SunDwarf/Jokusoramame

E1337Kat committed:
Merge https://github.com/SunDwarf/Jokusoramame

E1337Kat committed:
Merge https://github.com/SunDwarf/Jokusoramame

E1337Kat committed:
Merge https://github.com/SunDwarf/Jokusoramame

E1337Kat committed:
Changed columns to sort stocks by

E1337Kat committed:
Merge branch 'master' of https://github.com/E1337Kat/Jokusoramame

E1337Kat committed:
heck, I can't spell

E1337Kat committed:
I NEED TO LEARN TO REEAADDD

E1337Kat committed:
coding is hard on codine

E1337Kat committed:
god I'm dumb

E1337Kat committed:
(what == owned) or (what == amount) -_-

E1337Kat committed:
I am far too impatiant to read my own fricking code

E1337Kat committed:
richest sorts from big to low

E1337Kat committed:
assets sort high to low

E1337Kat committed:
git is annoying sometimes

E1337Kat committed:
Help should look better now

E1337Kat commented:
Everything seems good. dunno how it will perform on a large server....
